### PR TITLE
88: add entities methods create_many, delete, merge

### LINKED ISSUE #88: Script to use a CSV to bulk create, update, or delete Entities in an Entity List
We don't yet have a Entity List bulk upload from CSV API. We have one in the UI, but it's UI only and can't replace.

It'd be nice to have a script that given a CSV would:
1. create entities that don't exist in the Entity List
2. update entities that exist with data in the CSV
3. optionally, delete any entities that aren't in the CSV

I think we'd need to pass in the name of the column in the CSV that uniquely identifies each entity and also add a flag for whether or not to create new properties.

### PR DESCRIPTION
**Pull request by lindsay-stevens**
Closes #88

#### What has been done to verify that this works as intended?

Unit tests for the data processor method `_prep_data_for_merge`, and integration tests in `test_client.py`.

#### Why is this the best possible solution? Were any other approaches considered?

The original ticket called for an example script but 1) this seemed like functionality that would be widely useful enough to be a part of the library itself, and 2) it was somewhat complicated to implement and test (lots of options and edge cases) so as part of the library it should be easier to actively maintain.

Also, I added the use case to the existing example script `create_entities_from_submissions.py` because there was an almost complete overlap in the use case - it didn't involve updating data before, but you can do the same thing with `merge`. The tests contain examples of numerous other usage / call patterns.

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?

Provides some handy new ways to interact with Entities

#### Do we need any specific form for testing your changes? If so, please attach one.

No, the tests include all necessary data.

#### Before submitting this PR, please make sure you have:
- [x] included test cases for core behavior and edge cases in `tests`
- [x] run `python -m unittest` and verified all tests pass
- [x] run `ruff format pyodk tests` and `ruff check pyodk tests` to lint code
- [x] verified that any code or assets from external sources are properly credited in comments

### CODE REVIEW

**Comment by tobiasmcnulty** [comment]
@lindsay-stevens This looks great! Thanks for adding support for these methods directly in the library. I might not be able to test right away, but if real-world feedback would be helpful and is not too urgent, I can try to prioritize testing it and report back.

My only initial comment is that single-threaded deletes and updates can be quite slow (think 6 minutes instead of 6 seconds for ~5000 rows). It could be worth including a disclaimer in the docs, or adding support for multi-threaded deletes and updates. This is what we do in our code:

```python
def delete_entities(
    client,
    project_id,
    entity_list_name,
    uuids_to_delete,
):
    def delete_entity(uuid):
        delete_resp = client.delete(
            f"/projects/{project_id}/datasets/{entity_list_name}/entities/{uuid}",
        )
        maybe_fail_loudly_for_odk_resp(delete_resp)

    pool = ThreadPool(20)
    print(f"Deleting {len(uuids_to_delete)} UUIDs...")
    pool.map(delete_entity, uuids_to_delete)
    print(f"Done.")
```

**Comment by lindsay-stevens** [comment]
@tobiasmcnulty thanks! Testing would not be urgent, but appreciated if you have a chance. About the use cases:
- why is reducing the processing time important? Are entity updates needed that fast, or is it a cost thing etc?
- I can imagine needing to update 5k entities in one go, but what is the use case for a large batch of deletes?

Incorporating some kind of async might be out of scope for this PR. Since pyodk is a library, if it opens threadpools, and user app code then opens threadpools to invoke pyodk, we can end up with unexpectedly a lot of threads. Also a flood of requests to Central. I don't know if there are plans for it, but maybe Central could have a batch update and batch delete endpoint, like there is for create. Another approach could be make pyodk use aiohttp or similar, but that's a large refactor. With the current code, an approach could be to monkey-patch `EntityService.update` and `.delete` to use async, or make them no-ops and use the `merge` return values to run the updates/deletes with async.

**Comment by tobiasmcnulty** [comment]
We may have upwards of 3M entities to update (don't worry, not all in one list), so 6 minutes -> 6 seconds is a big improvement. I agree that long term it would be better to implement these changes server side, so take take the feedback with a grain of salt. Perhaps just a note in the docs so warn users before they can't figure out why `merge()` isn't returning, especially when `create_many()` is so fast?

File: pyodk/_endpoints/entities.py [hunk 1]
```diff
@@ -258,3 +366,190 @@ def get_table(
...
+                default_project_id=pid,
+                default_entity_list_name=eln,
+            )
+            for k in merge_actions.keys_difference:
+                elps.create(name=k)
+            merge_actions.final_keys = merge_actions.keys_union
+        else:
+            merge_actions.final_keys = merge_actions.keys_intersect
+        if len(merge_actions.to_insert) > 0:
+            insert_reshape = [
```
Comment (lognaturel) [inline pyodk/_endpoints/entities.py:532]: Basically I think this reshaping should to in `create_may` rather than here.

**Review by lognaturel — COMMENTED** [review]
I feel like the `create_many` and `merge` signatures should be more similar. I've made a few suggestions to that effect and am interested to hear what you think, @lindsay-stevens 

How about also adding a brief note about performance in the `merge` and `delete` doc comments, maybe also adding one to `update`? Something to the effect of "this method sends requests one at a time and may be slow for larger numbers of Entities. Consider making parallel calls and describing your use case on the ODK forum."

File: pyodk/_endpoints/entities.py [hunk 2]
```diff
@@ -144,6 +177,49 @@ def create(
...
+
+        Required keys in data: entities[].label, entities[].data, source.name.
+        Example of the required data format:
+
+        {
+            "entities": [
+                {"label": "Sydney", "data": {"state": "NSW", "postcode": "2000"}},
+                {"label": "Melbourne", "data": {"state": "VIC", "postcode": "3000"}},
+            ],
+            "source": {"name": "pyodk", "size": 1},
```
[the remark this answers is not in the document]
Comment (lindsay-stevens) [reply pyodk/_endpoints/entities.py:197]: It seemed like "size" was meant for CSV file size, or something generic. So I wasn't sure what pyodk could guess that to mean for `dict` input - number of rows? Number of row values? So I left it to the user to define.

File: pyodk/_endpoints/entities.py [hunk 3]
```diff
@@ -258,3 +366,190 @@ def get_table(
...
+        Creation is performed using the bulk creation endpoint. This method may be slow
+        for large quantities of updates or deletes, since for these operations each
+        change is a request in a loop. If this is a concern, set the parameters
+        `update_matched` and `delete_not_matched` to False and use the return value to
+        perform threaded or async requests for these data.
+
+        :param source_data: Data to use for updating Entities in Central.
+        :param entity_list_name: The name of the Entity List (Dataset) being referenced.
+        :param project_id: The id of the project this Entity belongs to.
+        :param match_keys: Dictionary keys common to source and target used to match rows.
```
[the remark this answers is not in the document]
Comment (lindsay-stevens) [reply pyodk/_endpoints/entities.py:493]: Do you mean that all input entity data must have a user-specified uuid? I thought the main use case was to pass in rows with a label + some other data for pyodk match/send. The `match_keys` parameter defaults to `label` (in `_prep_data_for_merge`).

File: pyodk/_endpoints/entities.py [hunk 4]
```diff
@@ -258,3 +366,190 @@ def get_table(
...
+        source_data: Iterable[Mapping[str, Any]],
+        entity_list_name: str | None = None,
+        project_id: int | None = None,
+        match_keys: Iterable[str] | None = None,
+        add_new_properties: bool = True,
+        update_matched: bool = True,
+        delete_not_matched: bool = False,
+        source_label_key: str = "label",
+        source_keys: Iterable[str] | None = None,
+        create_source: str = "pyodk",
```
[the remark this answers is not in the document]
Comment (lindsay-stevens) [reply pyodk/_endpoints/entities.py:475]: Like as in `create_source=f"pyodk v {__version__}"` or as in add a separate parameter (in which case where/how would it be used?)?

**Comment by lindsay-stevens** [comment]
>How about also adding a brief note about performance in the merge and delete doc comments, maybe also adding one to update? Something to the effect of "this method sends requests one at a time and may be slow for larger numbers of Entities. Consider making parallel calls and describing your use case on the ODK forum."

I added a note to `merge` in [d657b0f](https://github.com/getodk/pyodk/pull/93/commits/d657b0fbc4bf526968c196bd49e71f7b44db1e59), but I think adding this kind of statement to individual endpoint methods might be too much, because it would apply to basically all the pyodk endpoint methods (handling lots of forms, or submissions, etc).

File: pyodk/_endpoints/entities.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (lognaturel) [reply pyodk/_endpoints/entities.py:475]: Yes, in the same parameter.

File: pyodk/_endpoints/entities.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (lognaturel) [reply pyodk/_endpoints/entities.py:197]: Ah yes, you're right about `size`. We went back and forth on what it should represent and landed on it being optional. Can we make it an optional separate parameter for this method too, then? That feels easier for the end user to handle.

File: pyodk/_endpoints/entities.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (lognaturel) [reply pyodk/_endpoints/entities.py:493]: Sorry, I did lose the point there! I was thinking of a case where you'd download the existing data, make manual updates in a spreadsheet or other program, and then use `merge` to merge that back in. I think that will be a real usage pattern but you're right that matching from an external system using a natural id will likely be more common. So yes, I think you made the right call here, thanks for explaining. 👍

File: pyodk/_endpoints/entities.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (lindsay-stevens) [reply pyodk/_endpoints/entities.py]: Applied these changes in 72bb68e

File: pyodk/_endpoints/entities.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (lindsay-stevens) [reply pyodk/_endpoints/entities.py]: Applied these changes in 72bb68e

**Review by lognaturel — APPROVED** [review]
Thanks for making those updates!
